backend/src/org/commcare/session/SessionNavigator.java

```diff
@@ -95,36 +96,38 @@ private void handleGetDatum() {
 
     private void handleAutoSelect() {
         SessionDatum selectDatum = currentSession.getNeededDatum();
-        if (selectDatum.getLongDetail() == null) {
-            // No confirm detail defined for this entity select, so just set the case id right away
-            // and proceed
-            String autoSelectedCaseId = SessionDatum.getCaseIdFromReference(
-                    currentAutoSelectedCase, selectDatum, ec);
-            currentSession.setDatum(selectDatum.getDataId(), autoSelectedCaseId);
-            startNextSessionStep();
-        } else {
-            sendResponse(LAUNCH_CONFIRM_DETAIL);
+        if (selectDatum instanceof EntityDatum) {
+            EntityDatum entityDatum = (EntityDatum)selectDatum;
+            if (entityDatum.getLongDetail() == null) {
+                // No confirm detail defined for this entity select, so just set the case id right away
+                // and proceed
+                String autoSelectedCaseId = EntityDatum.getCaseIdFromReference(
+                        currentAutoSelectedCase, entityDatum, ec);
+                currentSession.setDatum(entityDatum.getDataId(), autoSelectedCaseId);
+                startNextSessionStep();
+            } else {
+                sendResponse(LAUNCH_CONFIRM_DETAIL);
+            }
         }
     }
 
     /**
-     *
      * Returns the auto-selected case for the next needed datum, if there should be one.
      * Returns null if auto selection is not enabled, or if there are multiple available cases
      * for the datum (and therefore auto-selection should not be used).
      */
     private TreeReference getAutoSelectedCase() {
```

Comment:
Not your code, but the whole handleGetDatum() autoSelect workflow seems janky. We get the autoSelectedCase, then  put the result in a class variable. then call handleAutoSelect and do the instanceof/cast with selectDatum both times. Maybe belongs in a separate refactor but could move everything besides the if (autoSelection == null) case inside an if (selectDatum instanceof EntityDatum) cast passing through the casted EntityDatum and TreeReference as needed